qkeras/quantizers.py

```diff
@@ -1641,57 +1641,47 @@ class quantized_tanh(BaseQuantizer):  # pylint: disable=invalid-name
 
   Attributes:
     bits: number of bits to perform quantization.
-    integer: number of bits to the left of the decimal point.
-    symmetric: if true, we will have the same number of values for positive
-               and negative numbers.
     use_stochastic_rounding: if true, we perform stochastic rounding.
+    use_real_tanh: if true, use the tanh function from Keras backend,
+      if false, use tanh that is defined as 2 * sigmoid(x) - 1
 
   Returns:
     Function that performs tanh + quantization to bits in the range -1.0 to 1.0.
   """
 
-  def __init__(self, bits=8, integer=0, symmetric=0,
-               use_stochastic_rounding=False):
+  def __init__(self, bits=8, use_stochastic_rounding=False, use_real_tanh=False):
     super(quantized_tanh, self).__init__()
     self.bits = bits
-    self.integer = integer
-    self.symmetric = symmetric
     self.use_stochastic_rounding = use_stochastic_rounding
+    self.use_real_tanh = use_real_tanh
 
   def __str__(self):
-    flags = [str(self.bits), str(self.integer)]
-    if self.symmetric or self.use_stochastic_rounding:
-      flags.append(str(int(self.symmetric)))
+    flags = [str(self.bits)]
     if self.use_stochastic_rounding:
       flags.append(str(int(self.use_stochastic_rounding)))
+    if self.use_real_tanh:
+      flags.append(str(int(self.use_real_tanh)))
     return "quantized_tanh(" + ",".join(flags) + ")"
 
   def __call__(self, x):
     non_sign_bits = self.bits - 1
-    m = pow(2, non_sign_bits)
-    m_i = pow(2, self.integer)
-    p = _sigmoid(x / m_i) * m
-    xq = m_i * tf.keras.backend.clip(
```

Comment:
With  `quantized_tanh(4,use_real_tanh=True)`, I'm seeing 17 unique values, which is more than 2^4=16. This means that there must be a bug since we have more levels than can be represented with 4 bits. I think the key missing part is the clipping function

Perhaps try something like this:
```
    if not self.use_real_tanh:
      p = _sigmoid(x) * m
      xq = tf.keras.backend.clip(2.0 * (_round_through(p, self.use_stochastic_rounding) / m) - 1.0,
                                       -1.0 + 1.0 / m,
                                       1.0 - 1.0 / m)
    else:
      p = K.tanh(x) * m
      xq = tf.keras.backend.clip((_round_through(p, self.use_stochastic_rounding) / m),
                                       -1.0 + 1.0 / m,
                                       1.0 - 1.0 / m)
```

Reply:
I just realized that even with the code I linked, the real `tanh` has more steps than the `_sigmoid`. This code should fix that issue such that both `use_real_tanh = True` and `use_real_tanh = False` have the same number of steps:

```
def __call__(self, x):
    non_sign_bits = self.bits - 1
    x = K.cast_to_floatx(x)
    m = K.cast_to_floatx(K.pow(2, non_sign_bits))
    p = K.tanh(x) if self.use_real_tanh else 2.0 * _sigmoid(x) - 1.0
    xq = tf.keras.backend.clip((_round_through(p * m, self.use_stochastic_rounding) / m),
                                -1.0 + 1.0 / m,
                                1.0 - 1.0 / m)
    return xq

Reply:
Thanks for the suggestion!